ops/model.py

```diff
@@ -938,18 +985,34 @@ def _is_mutable(self):
         return False
 
     def __setitem__(self, key: str, value: str):
-        if not self._is_mutable():
+        self._validate_write(key, value)
+        self._commit(key, value)
+        self._update(key, value)
+
+    def _commit(self, key, value):
+        self._backend.update_relation_data(self.relation.id, self._entity, key, value)
+
+    def _validate_write(self, key: str, value: str):
+        """Validate writing key:value to this databag.
+
+        1) that key: value is a valid str:str pair
+        2) that we have write access to this databag
+        """
+        if not self._is_writable():
             raise RelationDataError(
-                'cannot set relation data for {}'.format(self._entity.name))
+                '{} cannot write the application databag of {}. It could be that '
+                'the databag belongs to a remote unit/application, or that you '
+                'are attempting to write app data from a non-leader unit.'
+                ''.format(self._entity.name, self._backend.app_name))
```

Comment:
It looks like the only usage of `_is_writable` is here.  And so maybe it would be appropriate to just consolidate these two methods - and only have a `_vaidate_write` and a corresponding `_validate_read` that both do all the checks and raise nice, specific exceptions.